### PR TITLE
[NON-MODULAR] [BOUNTY] AI Quality Of Life Improvements

### PR DESCRIPTION
**Pull request by linnpap**
## About The Pull Request
Adds in several quality of life improvements for AI players. These include the ability to see and send LOOC messages through the AI eye, flavor text for the holopad when you are projecting, and also to see *mes that have been sent by mobs.

## Why It's Good For The Game
Should make RP a bit more viable, and not as much as a pain in the ass.

## Changelog
:cl:
qol: AI players can now see and send LOOC messages through the AI eye
qol: Adds flavor text to the holopad when someone is projecting
qol: AI players can now see actions that mobs have done
/:cl:

### CODE REVIEW

**Comment by SpaceVampire** [comment]
This sounds like the kind of PR that should really be done upstream.
Due to the non-modular nature of it, and I don't see any reason that TG wouldn't want it.

File: code/datums/emotes.dm [hunk 1]
```diff
@@ -127,6 +127,18 @@
 		if(ghost.stat == DEAD && ghost.client && user.client && (ghost.client.prefs.chat_toggles & CHAT_GHOSTSIGHT) && !(ghost in viewers(user_turf, null)))
 			ghost.show_message("<span class='emote'>[FOLLOW_LINK(ghost, user)] [dchatmsg]</span>")
 
+	//SKYRAT EDIT ADDITION BEGIN - AI QoL
```
Comment (FlamingLily) [inline code/datums/emotes.dm:130]: This could go upstream

File: code/game/machinery/hologram.dm [hunk 2]
```diff
@@ -201,6 +201,38 @@ Possible to do for anyone motivated enough:
 	else if(in_range(user, src) || isobserver(user))
 		. += span_notice("The status display reads: Current projection range: <b>[holo_range]</b> units.")
 
+	//SKYRAT EDIT ADDITION BEGIN - AI QoL
```
Comment (FlamingLily) [inline code/game/machinery/hologram.dm:204]: as could this

File: modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm [hunk 3]
```diff
@@ -39,6 +39,24 @@
 	mob.log_talk(msg,LOG_OOC, tag="LOOC")
 
 	var/list/heard = get_hearers_in_view(7, get_top_level_mob(src.mob))
+
+	//SKYRAT EDIT ADDITION BEGIN - AI QoL
```
Comment (FlamingLily) [inline modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm:43]: and this is modular

**Review by FlamingLily — COMMENTED** [review]
send it upstream, they'll like it

File: code/datums/emotes.dm [hunk 4]
```diff
@@ -127,6 +127,18 @@
...
+	for(var/mob/ai in GLOB.ai_list)
+		//ai.show_message("<span class='emote'>[!(ai.stat == DEAD)]</span>")
+		var/aiEye_turf
+		for(var/mob/camera/ai_eye/AI_eye as anything in GLOB.aiEyes)
+			aiEye_turf = get_turf(AI_eye)
+		if(!ai.client)
+			continue
+		if(!(ai.stat == DEAD) && ai.client && (get_dist(user_turf, aiEye_turf)<8))
```
Comment (ZephyrTFA) [inline code/datums/emotes.dm:138]: Dont hard code it like this. use world.view

File: code/game/machinery/hologram.dm [hunk 5]
```diff
@@ -201,6 +201,38 @@ Possible to do for anyone motivated enough:
...
+	var/obj/effect/overlay/holo_pad_hologram/holo
+	var/line
+	var/mob/living/silicon/ai/aiPlayer
```
Comment (ZephyrTFA) [inline code/game/machinery/hologram.dm:207]: Don't use camel case, use snake case

File: modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm [hunk 6]
```diff
@@ -39,6 +39,24 @@
...
+	var/aiAlreadyHeardIt = FALSE
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm]: snake_case

File: modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm [hunk 7]
```diff
@@ -39,6 +39,24 @@
...
+	//so the ai can post looc text
+	if(istype(src.mob,/mob/living/silicon/ai))
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm]: You don't need src here

File: modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm [hunk 8]
```diff
@@ -39,6 +39,24 @@
...
+		for(var/mob/camera/ai_eye/AI_eye as anything in GLOB.aiEyes)
+			heard = get_hearers_in_view(7, AI_eye)
+	//so the ai can see looc text
+	for(var/mob/camera/ai_eye/AI_eye as anything in GLOB.aiEyes)
+		for(var/mob/creature in heard) //checks if an ai has already heard it
+			if(istype(creature,/mob/living/silicon/ai))
+				aiAlreadyHeardIt = TRUE
+		if(aiAlreadyHeardIt == FALSE) //i dont wanna run an entire fuckin for loop if it turns out the ai can already see the text
+			for(var/mob/creature in heard)
+				if(creature==AI_eye)
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm]: you need whitespace around equality operators

File: modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm [hunk 9]
```diff
@@ -39,6 +39,24 @@
...
+	//SKYRAT EDIT ADDITION BEGIN - AI QoL
+	var/aiAlreadyHeardIt = FALSE
+	//so the ai can post looc text
+	if(istype(src.mob,/mob/living/silicon/ai))
+		for(var/mob/camera/ai_eye/AI_eye as anything in GLOB.aiEyes)
+			heard = get_hearers_in_view(7, AI_eye)
+	//so the ai can see looc text
+	for(var/mob/camera/ai_eye/AI_eye as anything in GLOB.aiEyes)
+		for(var/mob/creature in heard) //checks if an ai has already heard it
+			if(istype(creature,/mob/living/silicon/ai))
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm]: you need a space after colons

File: modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm [hunk 10]
```diff
@@ -39,6 +39,24 @@
...
 	var/list/heard = get_hearers_in_view(7, get_top_level_mob(src.mob))
+
+	//SKYRAT EDIT ADDITION BEGIN - AI QoL
+	var/aiAlreadyHeardIt = FALSE
+	//so the ai can post looc text
+	if(istype(src.mob,/mob/living/silicon/ai))
+		for(var/mob/camera/ai_eye/AI_eye as anything in GLOB.aiEyes)
+			heard = get_hearers_in_view(7, AI_eye)
+	//so the ai can see looc text
+	for(var/mob/camera/ai_eye/AI_eye as anything in GLOB.aiEyes)
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm]: dont capitalize like this, just ai_eye is fine

**Comment by ZephyrTFA** [comment]
> 
> 
> send it upstream, they'll like it

Upstream doesnt have LOOC

**Comment by SpaceVampire** [comment]
> Upstream doesnt have LOOC

Should have been more clear, because this PR is a number of changes in one.
Lily went ahead and pointed it out, but it seemed like the emote stuff could be done there.
Just in the spirit of trying to limit unnecessary non-modular additions.

Same for this PR as well, right? https://github.com/Skyrat-SS13/Skyrat-tg/pull/7066

**Comment by Gandalf2k15** [comment]
Looks good!

**Pull request merged**